### DISCUSSION TITLE
Information about tiling on custom dataset

### DISCUSSION THREAD

**Discussion by LorenzoF6** [open-edge-platform/anomalib#2073] [Q&A] [answered]
Hi, i'm new in the anomaly detection tasks and i'm trying to understand the mechanism behind tiling on the anomalib's library and i have several question to ask about it.
- all model support tiling or some of them ? which one ?
- if i have a custom dataset of 1024x2014 image resolution, tiling is needed to improve the model's performance?
thank you all for the response in advance

**Comment by blaz-r** [comment, marked as answer]
Check the [docs](https://anomalib.readthedocs.io/en/latest/markdown/guides/how_to/data/input_tiling.html) for tiling info.

But to answer your questions:
- models currently supported are: Padim, Patchcore, STFPM, Reverse Distillation.
- main goal of tiling is to enable processing of higher resolutions - this usually results in a better performance, however if you compare against the same resolution that is not tiled, then usually this is not the case.

**Reply by onkarkris** [reply]
Hi @blaz-r  I tried tiling ensemble on my custom dataset with Patchcore following -->
[https://github.com/openvinotoolkit/anomalib/pull/1226/files/dac985fac8cba8644c56cf9c5dd58adaeb651afd#diff-21aa73a1cc1f2758f5b4d44477721b0b6fd00b78976dd91ea019dd9d95624e52](url)

In my training set there are 80 images of 1024x1024 

I get an error --- ` File "/home/ubuntu/anomalib_tile/tools/tiled_ensemble/ensemble_functions.py", line 73, in __call__
    coll_batch["image"] = tiled_images[self.tile_index]
IndexError: index 8 is out of bounds for dimension 1 with size 8`



My config is -->
`dataset:
  name: piston
  format: mvtec
  path: 
  task: segmentation
  category: 
  train_batch_size: 1
  eval_batch_size: 1
  num_workers: 8
  image_size: 1100 # dimensions to which images are resized (mandatory)
  center_crop: 1024 # dimensions to which images are center-cropped after resizing (optional)
  normalization: imagenet # data distribution to which the images will be normalized: [none, imagenet]
  transform_config:
    train: null
    eval: null
  test_split_mode: from_dir # options: [from_dir, synthetic]
  test_split_ratio: 0.2 # fraction of train images held out testing (usage depends on test_split_mode)
  val_split_mode: same_as_test # options: [same_as_test, from_test, synthetic]
  val_split_ratio: 0.5 # fraction of train/test images held out for validation (usage depends on val_split_mode)
  tiling:
    apply: false
    tile_size: null
    stride: null
    remove_border_count: 0
    use_random_tiling: False
    random_tile_count: 16

model:
  name: patchcore
  backbone: wide_resnet50_2
  pre_trained: true
  layers:
    - layer2
    - layer3
  coreset_sampling_ratio: 0.1
  num_neighbors: 9
  normalization_method: min_max # options: [null, min_max, cdf]

metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
  threshold:
    method: adaptive #options: [adaptive, manual]
    manual_image: null
    manual_pixel: null

visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  log_images: True # log images to the available loggers (if any)
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]

project:
  seed: 0
  path: ./results

logging:
  logger: [] # options: [comet, tensorboard, wandb, csv] or combinations.
  log_graph: false # Logs the model graph to respective logger.

optimization:
  export_mode: null # options: onnx, openvino

# PL Trainer Args. Don't add extra parameter here.
trainer:
  enable_checkpointing: true
  default_root_dir: null
  gradient_clip_val: 0
  gradient_clip_algorithm: norm
  num_nodes: 1
  devices: 1
  enable_progress_bar: true
  overfit_batches: 0.0
  track_grad_norm: -1
  check_val_every_n_epoch: 1 # Don't validate before extracting features.
  fast_dev_run: false
  accumulate_grad_batches: 1
  max_epochs: 1
  min_epochs: null
  max_steps: -1
  min_steps: null
  max_time: null
  limit_train_batches: 1.0
  limit_val_batches: 1.0
  limit_test_batches: 1.0
  limit_predict_batches: 1.0
  val_check_interval: 1.0 # Don't validate before extracting features.
  log_every_n_steps: 50
  accelerator: auto # <"cpu", "gpu", "tpu", "ipu", "hpu", "auto">
  strategy: null
  sync_batchnorm: false
  precision: 32
  enable_model_summary: true
  num_sanity_val_steps: 0
  profiler: null
  benchmark: false
  deterministic: false
  reload_dataloaders_every_n_epochs: 0
  auto_lr_find: false
  replace_sampler_ddp: true
  detect_anomaly: false
  auto_scale_batch_size: false
  plugins: null
  move_metrics_to_cpu: false
  multiple_trainloader_mode: max_size_cycle
`

**Reply by blaz-r** [reply]
Hi @onkarkris, I'll have to look into this, but I suspect that the problem might be in center crop. I am now working on refactoring this code to work with v1, I'll also make sure it works there.

For the time being, can you try to use the input size of 1024 without center-crop.

**Reply by blaz-r** [reply]
I also noticed that you haven't shared the ens_config.yaml file that you are using. Can you also send that please?

**Reply by onkarkris** [reply]
@blaz-r thanks it would be great if you can find the issue! I will check and let you know if turning off crop solve the issue  

ens_config  -
`tiling:
  tile_size: 256
  stride: 256

predictions:
  storage: memory # how the predictions are stored, options: [memory, file_system, memory_downscaled]
  downscale_factor: 0.5 # factor by which data is downscaled in case of "memory_downscaled" storage

post_processing:
  normalization: image # on what level we normalize, options: [tile, image, none]
  smooth_joins:
    apply: True # if this is applied, area around tile joins are is smoothed
    sigma: 2 # sigma of gaussian filter used to smooth this area
    width: 0.1 # width factor, multiplied by tile dimension gives the region width around join which will be smoothed

# overrides metrics from model config
metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
  threshold:
    stage: image # stage at which we apply threshold, options: [tile, image]
    method: adaptive # options: [adaptive, manual]
    manual_image: null
    manual_pixel: null

# overrides visualization from model config
visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]
`

**Reply by blaz-r** [reply]
Hi, I tried running with 1024 and I was able to train. I'm pretty sure the problem is in center crop. I'll fix that for the v1 refactor, for now I recommend you don't use that.

**Reply by blaz-r** [reply]
I also managed to reproduce the issue, it is center-crop related, as the tiler takes image_size as input, but it should take center-crop if that is set. As I said, for the time being, avoid center-crop and just set the image size to the desired value - 1024 in your case.
